File: 210. Course Schedule II/findOrder.py
class Solution:
    def __init__(self):
        self.res = []
        self.adjacent = []
        self.visited = []

    def dfs(self, i):
        if self.visited[i] == 1:
            return False
        if self.visited[i] == 2:
            return True
        self.visited[i] = 1
        for j in self.adjacent[i]:
            if not self.dfs(j):
                return False
        self.visited[i] = 2
        self.res.append(i)
        return True

    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.visited = [0] * numCourses
        self.adjacent = [[] for i in range(numCourses)]
        for cur, pre in prerequisites:
            self.adjacent[cur].append(pre)
        for i in range(numCourses):
            if not self.dfs(i):
                return []
        return self.res

File: 210. Course Schedule II/findOrder.java
class Solution {
    private int k = 0;
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] in = new int[numCourses];
        int[] ans = new int[numCourses];
        List[] edges = new ArrayList[numCourses];
        for (int i = 0; i < numCourses; i++) {
            edges[i] = new ArrayList<Integer>();
        }
        for (int[] pre: prerequisites) {
            in[pre[0]] ++;
            edges[pre[1]].add(pre[0]);
        }
        for (int i = 0; i < numCourses; i++) {
            if (in[i] == 0) {
                ans[k ++] = i;
                in[i] = -1;
                topologicalSort(edges, edges[i], in, ans);
            }
        }
        return k == numCourses ? ans : new int[0];
    }

    public void topologicalSort(List[] edges, List<Integer> list, int[] in, int[] ans) {
        for (int course: list) {
            in[course] --;
            if (in[course] == 0) {
                ans[k ++] = course;
                in[course] = -1;
                topologicalSort(edges, edges[course], in, ans);
            }
        }
    }
}